#HW Wk 4. Data Cleaning
## Liam McKenzie

#Load in data and imports

In [1]:
import pandas as pd
import numpy as np
url = ("https://raw.githubusercontent.com/IndraniMandal/CSC310-S20/master/messy_covid19_southamerica%20-%20covid19_southamerica.csv")


In [2]:
c19df = pd.read_csv(url)
c19df.head()

ObservationDate Country_Region Province_State  Confirmed Deaths Recovered  \
0      2020-01-23         Mexico            NaN        0.0      0         0   
1      2020-01-23         Brazil            NaN        0.0      0         0   
2      2020-01-23       Colombia            NaN        0.0      0         0   
3      2020-02-26         Brazil            NaN        1.0      0         0   
4      2020-02-26         Brazil            NaN        1.0      0         0   

   Active  
0     0.0  
1     0.0  
2     0.0  
3     1.0  
4     1.0

#Question one:
###Which column has the maximum missing value? What % value is missing in that column? Should you keep this column or drop it? Why?


In [3]:
c19df.isnull().sum()

ObservationDate       0
Country_Region        0
Province_State     5989
Confirmed             1
Deaths                3
Recovered             2
Active               26
dtype: int64

In [4]:
maxMissPerc = c19df.isnull().sum().max()/len(c19df)*100
maxMissPerc

25.870410367170628

Province_State has the highest missing value with ≈ 25.87% of the data missing. Based on the scope of the data set, it should be safe to drop this column.

In [5]:
c19df = c19df.drop('Province_State', axis=1)
c19df.head()

ObservationDate Country_Region  Confirmed Deaths Recovered  Active
0      2020-01-23         Mexico        0.0      0         0     0.0
1      2020-01-23         Brazil        0.0      0         0     0.0
2      2020-01-23       Colombia        0.0      0         0     0.0
3      2020-02-26         Brazil        1.0      0         0     1.0
4      2020-02-26         Brazil        1.0      0         0     1.0

In [6]:
c19df.isnull().sum()

ObservationDate     0
Country_Region      0
Confirmed           1
Deaths              3
Recovered           2
Active             26
dtype: int64

#Question 2:
###There are some columns with only a few data points missing. Should you drop these columns too? Why or why not? How should you handle these specific data points(particular row)?

Now that Province_State is dropped, only a few null values exist. There is no reason to drop these values as they only have a small amount of NaNs and the columns provide useful information. Instead, they can be handled via imputation techniques like imputation via mean/median values.

Use imputation by mean value on 'Confirmed' because it is numerical data

In [7]:
c19df['Confirmed'].fillna(c19df['Confirmed'].mean(), inplace=True)


/tmp/ipython-input-3317063441.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  c19df['Confirmed'].fillna(c19df['Confirmed'].mean(), inplace=True)


In [8]:
c19df.Confirmed[c19df.Confirmed.isnull()]

Series([], Name: Confirmed, dtype: float64)

Use imputation by median on 'Deaths' because it holds categorical data.

In [9]:
c19df['Deaths'].fillna(0, inplace=True)

/tmp/ipython-input-2375528075.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  c19df['Deaths'].fillna(0, inplace=True)


In [10]:
c19df.Deaths[c19df.Deaths.isnull()]

Series([], Name: Deaths, dtype: object)

Check data type for 'Recovered' for best strategy for imputation.

In [11]:
c19df['Recovered']

0            0
1            0
2            0
3            0
4            0
         ...  
23145       64
23146     4667
23147     2275
23148     1755
23149    63346
Name: Recovered, Length: 23150, dtype: object

Because 'Recovered' is integers stored as dtype object, imputation by kNN is necessary. First, the data must be converted to type numeric

In [12]:
c19df['Recovered'] = pd.to_numeric(c19df['Recovered'], errors='coerce')

Group 'Recovered' and 'Active' into one value 'v'

In [13]:
v = c19df.loc[: ,'Recovered':'Active']

In [14]:
v

Recovered  Active
0            0.0     0.0
1            0.0     0.0
2            0.0     0.0
3            0.0     1.0
4            0.0     1.0
...          ...     ...
23145       64.0     0.0
23146     4667.0    67.0
23147     2275.0  2039.0
23148     1755.0   208.0
23149    63346.0  9568.0

[23150 rows x 2 columns]

The simple value imputer needs to be imported in order to perform kNN imputation on the data set for 'Recovered'

In [15]:
from sklearn.impute import KNNImputer

imputer = KNNImputer(n_neighbors=2)
clean_df = imputer.fit_transform(v)

In [16]:
#check imputation
clean_df

array([[    0.,     0.],
       [    0.,     0.],
       [    0.,     0.],
       ...,
       [ 2275.,  2039.],
       [ 1755.,   208.],
       [63346.,  9568.]])

The "new" data is not a part of the original data frame so it needs to be merged back in.

In [17]:
clean_df = pd.DataFrame(clean_df, columns= ['Recovered', 'Active'])
clean_df.head()

Recovered  Active
0        0.0     0.0
1        0.0     0.0
2        0.0     0.0
3        0.0     1.0
4        0.0     1.0

In [18]:
c19df.loc[ : ,'ObservationDate': 'Deaths']

ObservationDate                    Country_Region  Confirmed Deaths
0          2020-01-23                            Mexico        0.0      0
1          2020-01-23                            Brazil        0.0      0
2          2020-01-23                          Colombia        0.0      0
3          2020-02-26                            Brazil        1.0      0
4          2020-02-26                            Brazil        1.0      0
...               ...                               ...        ...    ...
23145      2020-09-28  Saint Vincent and the Grenadines       64.0      0
23146      2020-09-28                          Suriname     4836.0    102
23147      2020-09-28               Trinidad and Tobago     4386.0     72
23148      2020-09-28                           Uruguay     2010.0     47
23149      2020-09-28                         Venezuela    73528.0    614

[23150 rows x 4 columns]

Now that our kNN imputation is a dataframe we can concatinate it to the remaining dataset

In [19]:
transformed_df = pd.concat([c19df.loc[ : ,'ObservationDate': 'Deaths'],clean_df ], axis=1)
transformed_df

ObservationDate                    Country_Region  Confirmed Deaths  \
0          2020-01-23                            Mexico        0.0      0   
1          2020-01-23                            Brazil        0.0      0   
2          2020-01-23                          Colombia        0.0      0   
3          2020-02-26                            Brazil        1.0      0   
4          2020-02-26                            Brazil        1.0      0   
...               ...                               ...        ...    ...   
23145      2020-09-28  Saint Vincent and the Grenadines       64.0      0   
23146      2020-09-28                          Suriname     4836.0    102   
23147      2020-09-28               Trinidad and Tobago     4386.0     72   
23148      2020-09-28                           Uruguay     2010.0     47   
23149      2020-09-28                         Venezuela    73528.0    614   

       Recovered  Active  
0            0.0     0.0  
1            0.0     0.0  
2            0.0     0.0  
3            0.0     1.0  
4            0.0     1.0  
...          ...     ...  
23145       64.0     0.0  
23146     4667.0    67.0  
23147     2275.0  2039.0  
23148     1755.0   208.0  
23149    63346.0  9568.0  

[23150 rows x 6 columns]

#Question 3
### There are some data that has incorrect information like instead of a number they might have strings. How are you handling these data points(rows)?

First, check for remaining missing values or incorrect information in the dataset using our new df.

In [20]:
transformed_df.isnull().sum()

ObservationDate    0
Country_Region     0
Confirmed          0
Deaths             0
Recovered          0
Active             0
dtype: int64

In [21]:
c19df.isnull().sum()

ObservationDate     0
Country_Region      0
Confirmed           0
Deaths              0
Recovered           5
Active             26
dtype: int64

Since no missing values exist anymore, imputation techniques performed on the data set worked as intended.

#Question 4
### Display top ten countries with most number of confirmed cases

Group countries by confirmed cases

In [22]:
groupDF = transformed_df.groupby('Country_Region')['Confirmed'].max()

In [23]:
groupDF

Country_Region
Antigua and Barbuda                    101.0
Argentina                           723132.0
Bahamas                               3838.0
Barbados                               190.0
Belize                                1891.0
Bolivia                             134223.0
Brazil                              973142.0
Chile                               285939.0
Colombia                            265576.0
Costa Rica                           73714.0
Cuba                                  5483.0
Dominica                                30.0
Dominican Republic                  111666.0
Ecuador                             134965.0
El Salvador                          28809.0
French Guiana                           18.0
Grenada                                 24.0
Guadeloupe                              53.0
Guatemala                            90263.0
Guyana                                2787.0
Haiti                                 8740.0
Honduras                             75537.0
Jamaica                               6170.0
Martinique                              32.0
Mexico                              124314.0
Nicaragua                             5073.0
Panama                              111277.0
Paraguay                             39432.0
Peru                                367068.0
Saint Barthelemy                         3.0
Saint Kitts and Nevis                   19.0
Saint Lucia                             27.0
Saint Vincent and the Grenadines        64.0
Suriname                              4836.0
Trinidad and Tobago                   4386.0
Uruguay                               2010.0
Venezuela                            73528.0
Name: Confirmed, dtype: float64

Sort the grouped data by confirmed cases

In [24]:
sortedDF = groupDF.sort_values(ascending=False)

Display top 10 countries

In [25]:
sortedDF.head(10)

Country_Region
Brazil                973142.0
Argentina             723132.0
Peru                  367068.0
Chile                 285939.0
Colombia              265576.0
Ecuador               134965.0
Bolivia               134223.0
Mexico                124314.0
Dominican Republic    111666.0
Panama                111277.0
Name: Confirmed, dtype: float64

Export Cleaned dataset

In [26]:
from google.colab import files

transformed_df.to_csv('cleaned_data.csv', index=False)
files.download('cleaned_data.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>